### ISSUE TITLE
Tagged commits are ignored due to commitParseRegex

### ISSUE THREAD

**Issue by evanportwood** [release-flow/release-flow#9]
I am testing this package with my build pipeline and ran into an issue when attempting to generate a full release version from a release branch.

**What you expected to see, and what happened instead.**
I have a branch, releases/1.1, that has the last commit tagged with `v1.1`.  I expected the generated version to be 1.1.0.  Instead, the generated version is 1.1.0-beta.6.

**Your operating system and other environment information.**
I've encountered this issue on my computer and in Azure Pipelines.  The details are below.

Computer:
OS - Windows 10
Git - 2.30.0.windows.2

Azure Pipelines:
OS - Ubuntu 20.04 LTS
Git - 2.36.0

**As much information as possible, such as the command and configuration used.**
I'm running this command in the pipeline, `npx rfver -c src/rfconfig.yml -p -d`.

Here is the output:
```
Node version: v16.15.0
Command-line options:{"_":[],"c":"src/rfconfig.yml","config":"src/rfconfig.yml","p":true,"pretty":true,"d":true,"debug":true,"output":"json","o":"json","use-origin-branches":false,"useOriginBranches":false,"$0":"../../node_modules/.bin/rfver"}
Using options file 'src/rfconfig.yml'
Configuration options:{"trunkBranchName":"main","releaseBranchPrefix":"releases/","workingBranchPrefixes":["feature/","bugfix/","hotfix/","merge/"],"stripBranchPrefixFromLabel":true,"failOnUnknownPrefix":true,"strategy":{"kind":"SemVer","baseNumber":"1.1"}}
Running command git symbolic-ref -q HEAD
Command exit code 0
Stdout: refs/heads/releases/1.1

Stderr: 
Source branch detected: refs/heads/releases/1.1
Using SemVer strategy
Options:{"trunkBranchName":"main","releaseBranchPrefix":"releases/","workingBranchPrefixes":["feature/","bugfix/","hotfix/","merge/"],"stripBranchPrefixFromLabel":true,"failOnUnknownPrefix":true,"strategy":{"kind":"SemVer","baseNumber":"1.1"},"useOriginBranches":false}
Getting commit for ref 'HEAD'
Running command git show --no-patch --no-notes --pretty='%H %ct %d' HEAD
Command exit code 0
Stdout: 'ea20cf62ebd41ca8683e03d9c74857b47852ae51 1651585875  (HEAD -> releases/1.1, origin/releases/1.1)'

Stderr: 
Current commit: ea20cf62ebd41ca8683e03d9c74857b47852ae51
Trigger is commit to release branch '1.1'
Running command git branch --all --no-color
Command exit code 0
Stdout:   feature/muggy
  main
* releases/1.1
  remotes/origin/feature/muggy
  remotes/origin/main
  remotes/origin/releases/1.1

Stderr: 
Looking for branches with prefix releases/
Release branches: <<<
releases/1.1
<<<
Attempting to parse branch '1.1' as release branch
Running command git merge-base main releases/1.1
Command exit code 0
Stdout: 84e7c3a89ce891f58ecd8f58cd51ce5fdab0afc7

Stderr: 
Getting commit for ref '84e7c3a89ce891f58ecd8f58cd51ce5fdab0afc7'
Running command git show --no-patch --no-notes --pretty='%H %ct %d' 84e7c3a89ce891f58ecd8f58cd51ce5fdab0afc7
Command exit code 0
Stdout: '84e7c3a89ce891f58ecd8f58cd51ce5fdab0afc7 1651520624 '

Stderr: 
Running command git log --no-patch --boundary --pretty='%H %ct %d' --simplify-by-decoration 84e7c3a89ce891f58ecd8f58cd51ce5fdab0afc7..releases/1.1
Command exit code 0
Stdout: 'ea20cf62ebd41ca8683e03d9c74857b47852ae51 1651585875  (HEAD -> releases/1.1, origin/releases/1.1)'
'd43605f8a2f62cb69d1e45e19251f4e9accd62d3 1651538884  (tag: v1.1)'
'84e7c3a89ce891f58ecd8f58cd51ce5fdab0afc7 1651520624 '

Stderr: 
No release version tags found on branch releases/1.1
Base release number: '1.1.0'
Using version source from branch 'releases/1.1', version = 1.1.0
Running command git rev-list --count ea20cf62ebd41ca8683e03d9c74857b47852ae51 ^84e7c3a89ce891f58ecd8f58cd51ce5fdab0afc7
Command exit code 0
Stdout: 6

Stderr: 
{
  "major": 1,
  "minor": 1,
  "patch": 0,
  "preReleaseLabel": "beta",
  "sha": "ea20cf62ebd41ca8683e03d9c74857b47852ae51",
  "buildType": "beta",
  "branchName": "releases/1.1",
  "commitDate": "2022-05-03T13:51:15.000Z",
  "commitsSinceVersionSource": 6,
  "versionSourceSha": "84e7c3a89ce891f58ecd8f58cd51ce5fdab0afc7",
  "majorMinor": "1.1",
  "majorMinorPatch": "1.1.0",
  "shortSha": "ea20cf6",
  "semVer": "1.1.0-beta.6"
}
```

I cloned the repo and updated the `commitParseRegex` to this value, `/^'?(?<sha>[0-9a-f]{40})\s(?<date>\d+)\s*(?:\((?<tags>.*)\))?'?$/;` and all tests passed.

I can submit a PR if you'd like.  Thanks for the great work!  I was excited to come across this package!

**Comment by richtea** [comment]
Hi Evan, I'm currently on a ferry returning from holiday, I will take a closer look on my return tomorrow. From what I can see, though, the tag should be the full version number (major / minor / patch), which may be why the regex isn't picking up the tag. I will take a deeper look when I am on my computer.

**Comment by evanportwood** [comment]
Thank you for the feedback.  I'll try your recommendation.

Thanks,
Evan

**Comment by evanportwood** [comment]
In my testing, I have created a new release branch, `releases/1.2`.  I tagged the HEAD commit with `v1.2.1`, ran my pipeline, and the generated version is `1.2.2-beta.0`.

I assume I should tag a commit in the release branch since we don't really need to tag commits on the main branch in Release Flow.  Please advise if that is incorrect.

Here is the output from the `npx rfver -c src/rfconfig.yml -p -d` command.
```
Node version: v16.15.0
Command-line options:{"_":[],"c":"src/rfconfig.yml","config":"src/rfconfig.yml","p":true,"pretty":true,"d":true,"debug":true,"output":"json","o":"json","use-origin-branches":false,"useOriginBranches":false,"$0":"../../node_modules/.bin/rfver"}
Using options file 'src/rfconfig.yml'
Configuration options:{"trunkBranchName":"main","releaseBranchPrefix":"releases/","workingBranchPrefixes":["feature/","hotfix/","pull/"],"stripBranchPrefixFromLabel":true,"failOnUnknownPrefix":true,"strategy":{"kind":"SemVer","baseNumber":"1.1"}}
Running command git symbolic-ref -q HEAD
Command exit code 0
Stdout: refs/heads/releases/1.2

Stderr: 
Source branch detected: refs/heads/releases/1.2
Using SemVer strategy
Options:{"trunkBranchName":"main","releaseBranchPrefix":"releases/","workingBranchPrefixes":["feature/","hotfix/","pull/"],"stripBranchPrefixFromLabel":true,"failOnUnknownPrefix":true,"strategy":{"kind":"SemVer","baseNumber":"1.1"},"useOriginBranches":false}
Getting commit for ref 'HEAD'
Running command git show --no-patch --no-notes --pretty='%H %ct %d' HEAD
Command exit code 0
Stdout: '92a520722dcad0d48115879d723c51fcbb6292ca 1651598091  (HEAD -> releases/1.2, tag: v1.2.1, origin/releases/1.2)'

Stderr: 
Current commit: 92a520722dcad0d48115879d723c51fcbb6292ca
Trigger is commit to release branch '1.2'
Running command git branch --all --no-color
Command exit code 0
Stdout:   main
  releases/1.1
* releases/1.2
  remotes/origin/main
  remotes/origin/releases/1.1
  remotes/origin/releases/1.2

Stderr: 
Looking for branches with prefix releases/
Release branches: <<<
releases/1.1
releases/1.2
<<<
Attempting to parse branch '1.1' as release branch
Running command git merge-base main releases/1.1
Command exit code 0
Stdout: 84e7c3a89ce891f58ecd8f58cd51ce5fdab0afc7

Stderr: 
Getting commit for ref '84e7c3a89ce891f58ecd8f58cd51ce5fdab0afc7'
Running command git show --no-patch --no-notes --pretty='%H %ct %d' 84e7c3a89ce891f58ecd8f58cd51ce5fdab0afc7
Command exit code 0
Stdout: '84e7c3a89ce891f58ecd8f58cd51ce5fdab0afc7 1651520624 '

Stderr: 
Attempting to parse branch '1.2' as release branch
Running command git merge-base main releases/1.2
Command exit code 0
Stdout: c96f520a3388da556444918493bdcefdf258785e

Stderr: 
Getting commit for ref 'c96f520a3388da556444918493bdcefdf258785e'
Running command git show --no-patch --no-notes --pretty='%H %ct %d' c96f520a3388da556444918493bdcefdf258785e
Command exit code 0
Stdout: 'c96f520a3388da556444918493bdcefdf258785e 1651597275 '

Stderr: 
Running command git log --no-patch --boundary --pretty='%H %ct %d' --simplify-by-decoration c96f520a3388da556444918493bdcefdf258785e..releases/1.2
Command exit code 0
Stdout: '92a520722dcad0d48115879d723c51fcbb6292ca 1651598091  (HEAD -> releases/1.2, tag: v1.2.1, origin/releases/1.2)'
'c96f520a3388da556444918493bdcefdf258785e 1651597275 '

Stderr: 
Base release number: '1.1.0'
Using version source from release tag 'v1.2.1', version = 1.2.1
Running command git rev-list --count 92a520722dcad0d48115879d723c51fcbb6292ca ^92a520722dcad0d48115879d723c51fcbb6292ca
Command exit code 0
Stdout: 0

Stderr: 
{
  "major": 1,
  "minor": 2,
  "patch": 2,
  "preReleaseLabel": "beta",
  "sha": "92a520722dcad0d48115879d723c51fcbb6292ca",
  "buildType": "beta",
  "branchName": "releases/1.2",
  "commitDate": "2022-05-03T17:14:51.000Z",
  "commitsSinceVersionSource": 0,
  "versionSourceSha": "92a520722dcad0d48115879d723c51fcbb6292ca",
  "majorMinor": "1.2",
  "majorMinorPatch": "1.2.2",
  "shortSha": "92a5207",
  "semVer": "1.2.2-beta.0"
}
```

**Comment by richtea** [comment]
> I assume I should tag a commit in the release branch since we don't really need to tag commits on the main branch in Release Flow.

Yes, that's right. I will update the documentation to explain that.

I think the reason for the unexpected release number is this:
> `Trigger is commit to release branch '1.2'`

release-flow expects the release build to be triggered by a tag, not a commit (the act of tagging is seen as marking the tagged commit as the final release version). See [this code](https://github.com/release-flow/release-flow/blob/d504cc2757644f34eb4ae0b0b381ea9ed2c97096/src/BuildVersionCalculator.ts#L301) for how it works. Again, I need to update the docs to make this clear.

**Comment by evanportwood** [comment]
My pipeline has triggers for any `releases/*` branch and any `v1.*` tag.  I should clarify that I'm using Azure Pipelines and not GitHub Actions, in case that could be a factor.

Here are the triggers from my pipeline yaml.
```
trigger:
  branches:
    include:
    - main
    - releases/*
  tags:
    include:
    - v1.*
```

The reason for the `releases/*` trigger is that I want to build and deploy a beta image when the release branch is cut and after any hot-fixes are cherry-picked into a release branch.  Once validation is complete and we're ready to release, I will tag the final commit to trigger the final build and release that will be deployed to production.

During my testing, I tagged the final commit as v1.2.0 and the pipeline still failed.  It doesn't recognize `refs/heads/tags/v1.2.0` which is what Azure Pipelines is returning when `git symbolic-ref -q HEAD` is executed.  I'm not sure why it isn't `refs/tags/v1.2.0`, but I think updating the regex on [line 303](https://github.com/release-flow/release-flow/blob/d504cc2757644f34eb4ae0b0b381ea9ed2c97096/src/BuildVersionCalculator.ts#L303) to `^refs\/(heads\/)?tags\/(?<Version>v.*)$` will work.

I tested this in regex101.com
<img width="422" alt="image" src="https://user-images.githubusercontent.com/96517/166709173-c7bed547-d8c0-4747-bad7-7ba77628d5fc.png">


Here is the error message from the failed pipeline run that was triggered by the addition of the v1.2.0 tag.
```
Node version: v16.15.0
Command-line options:{"_":[],"c":"src/rfconfig.yml","config":"src/rfconfig.yml","p":true,"pretty":true,"d":true,"debug":true,"output":"json","o":"json","use-origin-branches":false,"useOriginBranches":false,"$0":"../../node_modules/.bin/rfver"}
Using options file 'src/rfconfig.yml'
Configuration options:{"trunkBranchName":"main","releaseBranchPrefix":"releases/","workingBranchPrefixes":["feature/","hotfix/","pull/"],"stripBranchPrefixFromLabel":true,"failOnUnknownPrefix":true,"strategy":{"kind":"SemVer","baseNumber":"1.1"}}
Running command git symbolic-ref -q HEAD
Command exit code 0
Stdout: refs/heads/tags/v1.2.0

Stderr: 
Source branch detected: refs/heads/tags/v1.2.0
Using SemVer strategy
Options:{"trunkBranchName":"main","releaseBranchPrefix":"releases/","workingBranchPrefixes":["feature/","hotfix/","pull/"],"stripBranchPrefixFromLabel":true,"failOnUnknownPrefix":true,"strategy":{"kind":"SemVer","baseNumber":"1.1"},"useOriginBranches":false}
Getting commit for ref 'HEAD'
Running command git show --no-patch --no-notes --pretty='%H %ct %d' HEAD
Command exit code 0
Stdout: '92a520722dcad0d48115879d723c51fcbb6292ca 1651598091  (HEAD -> tags/v1.2.0, tag: v1.2.0, origin/releases/1.2, releases/1.2)'

Stderr: 
Current commit: 92a520722dcad0d48115879d723c51fcbb6292ca
Trigger is other branch 'refs/heads/tags/v1.2.0'
ERROR: Unsupported source ref 'refs/heads/tags/v1.2.0'
ERROR: Unsupported source ref 'refs/heads/tags/v1.2.0'
##[error]Bash exited with code '255'.
Finishing: Generate ReleaseFlow Version
```

**Comment by richtea** [comment]
I see - apologies for being a bit slow!

Strange that the git ref is `refs/heads/tags/vXXX` rather than `refs/tags/vXXX`. I use Azure Pipelines too, but in my pipelines the ref is just `refs/tags/vXXX` as you'd expect. The only difference is that, in my YAML pipelines, the triggers are all configured with the  `branches` keyword, even the tags, like this:

```yaml
trigger:
  branches:
    include:
    - main
    - release/*
    - refs/tags/v*
```

I assume this must affect how AzP checks out the repo.

> I should clarify that I'm using Azure Pipelines and not GitHub Actions, in case that could be a factor.

The code was originally written while working with AzP, so that should be fine. Is your repo also in ADO, or is it in GitHub? I guess that might make a difference :confused:

> The reason for the releases/* trigger is that I want to build and deploy a beta image when the release branch is cut and after any hot-fixes are cherry-picked into a release branch. Once validation is complete and we're ready to release, I will tag the final commit to trigger the final build and release that will be deployed to production.

This is exactly how my pipelines are configured too. The only difference I can see is the trigger syntax.

I'd be very grateful for a PR. Would you also mind adding a test to demonstrate that the alternative ref format works? [This test](https://github.com/release-flow/release-flow/blob/d504cc2757644f34eb4ae0b0b381ea9ed2c97096/src/BuildVersionCalculator.spec.ts#L200) should be a good template.

**Comment by evanportwood** [comment]
The code is also hosted in ADO.

I completely missed the option to use `refs/tags/v*` branch pattern in the trigger definition.  I tried that but it still failed.  

The interesting bit is that it shows `refs/tags/` in the checkout task, `Starting: Checkout poc-microservice-aks.gitversion@refs/tags/v1.2.0 to s`.

Here is the error message from rfver:
```
Node version: v16.15.0
Command-line options:{"_":[],"c":"src/rfconfig.yml","config":"src/rfconfig.yml","p":true,"pretty":true,"d":true,"debug":true,"output":"json","o":"json","use-origin-branches":false,"useOriginBranches":false,"$0":"../../node_modules/.bin/rfver"}
Using options file 'src/rfconfig.yml'
Configuration options:{"trunkBranchName":"main","releaseBranchPrefix":"releases/","workingBranchPrefixes":["feature/","hotfix/","pull/"],"stripBranchPrefixFromLabel":true,"failOnUnknownPrefix":true,"strategy":{"kind":"SemVer","baseNumber":"1.1"}}
Running command git symbolic-ref -q HEAD
Command exit code 0
Stdout: refs/heads/tags/v1.2.0

Stderr: 
Source branch detected: refs/heads/tags/v1.2.0
Using SemVer strategy
Options:{"trunkBranchName":"main","releaseBranchPrefix":"releases/","workingBranchPrefixes":["feature/","hotfix/","pull/"],"stripBranchPrefixFromLabel":true,"failOnUnknownPrefix":true,"strategy":{"kind":"SemVer","baseNumber":"1.1"},"useOriginBranches":false}
Getting commit for ref 'HEAD'
Running command git show --no-patch --no-notes --pretty='%H %ct %d' HEAD
Command exit code 0
Stdout: '8739c1365b4a996c19caa4120edfd996f4f04cfc 1651687748  (HEAD -> tags/v1.2.0, tag: v1.2.0, origin/releases/1.2, releases/1.2)'

Stderr: 
Current commit: 8739c1365b4a996c19caa4120edfd996f4f04cfc
Trigger is other branch 'refs/heads/tags/v1.2.0'
ERROR: Unsupported source ref 'refs/heads/tags/v1.2.0'
ERROR: Unsupported source ref 'refs/heads/tags/v1.2.0'
##[error]Bash exited with code '255'.
Finishing: Generate ReleaseFlow Version
```

I'll submit a PR and include the test.  Thank you for your time!

**Comment by evanportwood** [comment]
I apologize for the delay, I was swamped with other work.  I have my change ready, but before I submit my PR, I wanted to know if I could include an updated `package-lock.json`?

When I ran `npm init`, there were a lot of npm audit warnings and was able to resolve them all by running `npm audit fix`.

I can omit this change if you'd prefer.

**Comment by richtea** [comment]
Please include the npm updates. I usually work through the Dependabot warnings every few weeks, but if you've made the changes I'm happy to incorporate them.

**Comment by evanportwood** [comment]
@richtea PR #11 has been created.

**Comment by richtea** [comment]
Thanks @evanportwood , I'll merge it tomorrow, and update the [extension](https://marketplace.visualstudio.com/items?itemName=release-flow-dev.release-flow-tools) to include the changes.

**Comment by richtea** [comment]
Thanks again @evanportwood, I've released v0.1.2 of the [NPM Package](https://www.npmjs.com/package/@release-flow/release-flow). I'll update the extension next in case you want to use it.

**Comment by evanportwood** [comment]
Thank you, Rich!  I really appreciate your work on this package.

Regards,
Evan

**Comment by evanportwood** [comment]
@richtea, I saw that https://github.com/release-flow/release-flow/actions/runs/2323910211 failed.  ~~I'll submit another PR with the correct syntax.~~

I just saw you already updated the regex.  That's much cleaner.  Thanks!

**Comment by richtea** [comment]
@evanportwood the automated checks failed due to linting, and when I looked into it I noticed a possible optimization.

Now that you are no longer a first-time contributor, I hope that the checks would run automatically on any future PR you might submit, so you'd get rapid feedback. But I'm new to the GH security model, that's just my understanding from reading.

Thanks again for being the first contributor :thumbsup:

**Comment by evanportwood** [comment]
@richtea thanks again for your assistance.  I have a follow-up question for you.  

#### How are you accounting for the detached HEAD state in Azure Pipelines?

I was testing the release-flow package alongside GitVersion and I didn't realize that GitVersion was masking a problem until I disabled that task in my pipeline.  As soon as I disabled it, the release-flow package fails on the `git symbolic-ref -q HEAD` command because Azure Pipelines checks out the commit rather than the branch.

The error message is: `ERROR: No source ref specified, and unable to determine (detached head)`

I reviewed GitVersion's [source code](https://github.com/GitTools/GitVersion/blob/acd21a9e693a70bb61869ec7ff957d662f262d97/src/GitVersion.Core/Core/GitPreparer.cs#L152) and found that they use a local branch to avoid a detached state.

**Comment by richtea** [comment]
Have you tried using the `--source-ref` option? The [extension code](https://github.com/release-flow/azure-pipelines-extension/blob/b9f89056667de3700606b9613dad085b44b8921e/tasks/ReleaseFlowGitVersionV0/index.ts#L71) set the source branch based on a specified parameter that defaults to `$(Build.SourceBranch)` - maybe try this:

```shell
npx rfver -c src/rfconfig.yml -p -d -s $(Build.SourceBranch)
```

`release-flow` will also produce better results if you specify the target branch, so that it can infer the intent behind the build. It assumes you are intending to merge to `main` (or your chosen trunk branch) but that's not the case if you're cherry-picking onto a release branch. There's a `--target-branch` option that you can use, [in the extension](https://github.com/release-flow/azure-pipelines-extension/blob/b9f89056667de3700606b9613dad085b44b8921e/tasks/ReleaseFlowGitVersionV0/task.json#L45) this defaults to `$(System.PullRequest.TargetBranch)`. Bear in mind that this will be empty if this isn't a PR build, so you might need to use quotes around the CLI option to allow an empty string.

**Comment by richtea** [comment]
I will add that you might want to use the `--use-origin-branches` [option](https://github.com/release-flow/release-flow/blob/454df624050675b5787340751c7313aed3c7d949/src/cli.ts#L132-L136) in case the local repo doesn't create a local branch for the release branches. The [extension code](https://github.com/release-flow/azure-pipelines-extension/blob/b9f89056667de3700606b9613dad085b44b8921e/tasks/ReleaseFlowGitVersionV0/index.ts#L100) does this, so I assume it's important when building on Azure DevOps. (Apologies for being a bit vague, it's been a little while since I was struggling with all these problems myself, mercifully :relieved:)
